Add extensibility to Credential Response (#391)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

If the Credential Response is not encrypted, the media type of the response MUST be set to `application/json`.

The following parameters are used in the JSON-encoded Credential Response body:

* `credentials`: OPTIONAL. Contains an array of one or more issued Credentials. It MUST NOT be used if the `transaction_id` parameter is present. The elements of the array MUST be objects. This specification defines the following parameters to be used inside this object:
   * `credential`: REQUIRED. Contains one issued Credential. It MAY be a string or an object, depending on the Credential Format. See Appendix A for the Credential Format-specific encoding requirements.

> * `transaction_id`: OPTIONAL. String identifying a Deferred Issuance transaction. This claim is contained in the response if the Credential Issuer cannot immediately issue the Credential. The value is subsequently used to obtain the respective Credential with the Deferred Credential Endpoint (see (#deferred-credential-issuance)). It MUST not be used if the `credentials` parameter is present. It MUST be invalidated after the Credential for which it was meant has been obtained by the Wallet.

Reviewer: Not everything in JSON is a claim. Most things, in fact, are not. 

```suggestion
* `transaction_id`: OPTIONAL. String identifying a Deferred Issuance transaction. This parameter is contained in the response if the Credential Issuer cannot immediately issue the Credential. The value is subsequently used to obtain the respective Credential with the Deferred Credential Endpoint (see (#deferred-credential-issuance)). It MUST not be used if the `credentials` parameter is present. It MUST be invalidated after the Credential for which it was meant has been obtained by the Wallet.
```

The following parameters are used in the JSON-encoded Credential Response body:

> * `credentials`: OPTIONAL. Contains an array of one or more issued Credentials. It MUST NOT be used if the `transaction_id` parameter is present. The elements of the array MUST be objects. The following parameters are used inside this object (other parameters MAY also be included):

Reviewer 2: ```suggestion
* `credentials`: OPTIONAL. Contains an array of one or more issued Credentials. It MUST NOT be used if the `transaction_id` parameter is present. The elements of the array MUST be objects. This specification defines the following parameters to be used inside this object:
```

I think PR #382 should cover the MAY part `Additional Credential Response parameters MAY be defined and used.`

Author: My thought was that it makes sense to state it here explicitly because this is a nested object, not a direct Credential Response parameter and we are already expecting more parameters to exist here for things like key derivation. Would you be fine with keeping it like this?
```suggestion
* `credentials`: OPTIONAL. Contains an array of one or more issued Credentials. It MUST NOT be used if the `transaction_id` parameter is present. The elements of the array MUST be objects. This specification defines the following parameters to be used inside this object: (other parameters MAY also be included):
```

   * `credential`: REQUIRED. Contains one issued Credential. It MAY be a string or an object, depending on the Credential Format. See Appendix A for the Credential Format-specific encoding requirements.

>    * `notification_id`: OPTIONAL. String identifying an issued Credential that the Wallet includes in the Notification Request as defined in (#notification).

Reviewer 3: Shouldn't we have a single `notification_id` for the whole response?

Reviewer 4: I've been struggling with the same question for a few days:

A `notification_id` for each credential in the batch is cleaner/simpler, as we don't need to worry about partial states (i.e. the question of what notification to make if you receive 10 credentials and only manage to store 5 of them).

However it's kindof sucky to make 10 notification http calls for a batch of 10 credentials.

I'm probably veering towards cleaner/simpler and ignoring the 'kind of sucky' part.

Reviewer 3: We may change the notification request to an array as well?

Reviewer 4: I think that could make a lot of sense.

Reviewer 5: FWIW, I think that `notification_id` should point to a single credential not in a batch.

For instance, an Issuer can react to a notification event carrying a `credential_deleted` or `credential_failure` to invalidate the credential from a status list.

We can avoid multiple posts for notifications by using an array of `notification_id` per possible event.

```json
{
  "credential_accepted" : [ "notId#1"],
  "credential_failure" : ["notid#4"],
  "credential_deteled": [ ]
}
```

Something like a summary notification

Reviewer 6: I think it might be useful to move it in because of `credential_deleted` which is specific to one issued credential although when I think about it, `notification_id` overlaps a bit with the concept of credential version IDs or data set IDs but this is out of scope of this PR.

Reviewer 3: Why would a wallet delete a single instance rather than the whole batch or the same Credential Dataset?

Reviewer 2: maybe because the user has used that one instance? but not sure that information needs to be communicated to the issuer real-time, since that would allow the issuer to get more information about when credential is used

* `c_nonce`: OPTIONAL. String containing a nonce to be used to create a proof of possession of key material when requesting a Credential (see (#credential-request)). When received, the Wallet MUST use this nonce value for its subsequent Credential Requests until the Credential Issuer provides a fresh nonce.
* `c_nonce_expires_in`: OPTIONAL. Number denoting the lifetime in seconds of the `c_nonce`.

> * `notification_id`: OPTIONAL. String identifying an issuance flow in which one or more Credentials were issued. It MUST be included in the Notification Request as defined in (#notification). It MUST not be used if the `credentials` parameter is not present.

Reviewer 2: ```suggestion
* `notification_id`: OPTIONAL. String identifying a Credential Request in which one or more Credentials were issued. It MUST be included in the Notification Request as defined in (#notification). It MUST not be used if the `credentials` parameter is not present.
```

can we please not use a term "issuance flow"? it could mean so many things.

Cache-Control: no-store

{
  "credentials": [
    {
      "credential": "LUpixVCWJk0eOt4CXQe1NXK....WZwmhmn9OQp6YxX0a2L"
    }
  ],
  "c_nonce": "fGFF7UkhLa",
  "c_nonce_expires_in": 86400
}
```

> Below is a non-normative example of a Credential Response in an immediate issuance flow for multiple Credential instances in JWT VC format (JSON encoded) with the additional `notification_id` parameter:

Reviewer 2: ```suggestion
Below is a non-normative example of a Credential Response in an immediate issuance flow for multiple Credential instances in JWT VC format (JSON encoded) with an additional `notification_id` parameter:
```

## Deferred Credential Response {#deferred-credential-response}

> The Deferred Credential Response uses the `credentials` and `notification_id` parameters as defined in (#credential-response).

Reviewer 2: ```suggestion
The Deferred Credential Response uses the `credentials` parameter as defined in (#credential-response).
```

The Deferred Credential Response MUST be sent using the `application/json` media type.

The following is a non-normative example of a Deferred Credential Response:

```
HTTP/1.1 200 OK
Content-Type: application/json

{
  "credentials": [
    {
      "credential": "LUpixVCWJk0eOt4CXQe1NXK....WZwmhmn9OQp6YxX0a2L"
    },
    {
      "credential": "YXNkZnNhZGZkamZqZGFza23....29tZTIzMjMyMzIzMjMy"
    }

>   ],
>   "notification_id": "3fwe98js"

Reviewer 2: ```suggestion
  ]
```

# Notification Endpoint {#notification-endpoint}

> This endpoint is used by the Wallet to notify the Credential Issuer of certain events for issued Credentials. These events enable the Credential Issuer to take subsequent actions after issuance. The Credential Issuer needs to return a `notification_id` parameter in the Credential Response or Deferred Credential Response for the Wallet to be able to use this endpoint. Support for this endpoint is OPTIONAL. The Issuer cannot assume that a notification will be sent for every issued Credential since the use of this Endpoint is not mandatory for the Wallet.

Reviewer 2: ```suggestion
This endpoint is used by the Wallet to notify the Credential Issuer of certain events for issued Credentials. These events enable the Credential Issuer to take subsequent actions after issuance. The Credential Issuer needs to return one `notification_id` parameter per Credential Response or Deferred Credential Response for the Wallet to be able to use this endpoint. Support for this endpoint is OPTIONAL. The Issuer cannot assume that a notification will be sent for every issued Credential since the use of this Endpoint is not mandatory for the Wallet.
```

## Notification Request {#notification}

> The Wallet sends an HTTP POST request to the Notification Endpoint with the following parameters in the entity-body and using the `application/json` media type. If the Wallet supports the Notification Endpoint, the Wallet MAY send one or more Notification Requests per Credential issuance flow.

Reviewer 5: I think the "notification_id` value received" is indeed better. 

I wallet may call multiple times the credential endpoint, using the same `access_token`. Thus, it can receive more than one `notification_id`.

* `notification_id`: REQUIRED. String received in the Credential Response or Deferred Credential Response identifying an issuance flow that contained one or more Credentials with the same Credential Configuration and Credential Dataset.

> * `event`: REQUIRED. Type of the notification event. It MUST be a case sensitive string whose value is either `credential_accepted`, `credential_failure`, or `credential_deleted`. `credential_accepted` is to be used when the Credentials were successfully stored in the Wallet, with or without user action. `credential_deleted` is to be used when the unsuccessful Credential issuance was caused by a user action. In all other unsuccessful cases, `credential_failure` is to be used. Partial errors in an Credential issuance flow (e.g., one of the Credentials could not be stored) SHOULD be treated as the overall issuance flow failing.

Reviewer 5: I believe that either we should drop the `credential_` prefix or change it to `credentials_`. 

Perhaps, though, this could be addressed in a separate issue/PR.

Reviewer 2: ```suggestion
* `event`: REQUIRED. Type of the notification event. It MUST be a case sensitive string whose value is either `credential_accepted`, `credential_failure`, or `credential_deleted`. `credential_accepted` is to be used when the Credentials were successfully stored in the Wallet, with or without user action. `credential_deleted` is to be used when the unsuccessful Credential issuance was caused by a user action. In all other unsuccessful cases, `credential_failure` is to be used. Partial errors during batch credential issuance (e.g., one of the Credentials in the batch could not be stored) MUST be treated as the overall issuance flow failing.
```

i am really not sure what "an Credential issuance flow" is. I think what we mean here is `Partial errors during batch credential issuance`?
also, I think it should be a MUST. SHOULD here is not really helpful.
i am pretty neutral on `credential_` vs `credentials_`. it's a batch of credentials, but of the same credential type

## Notification Request {#notification}

The Wallet sends an HTTP POST request to the Notification Endpoint with the following parameters in the entity-body and using the `application/json` media type. If the Wallet supports the Notification Endpoint, the Wallet MAY send one or more Notification Requests per `notification_id` value received.
* `notification_id`: REQUIRED. String received in the Credential Response or Deferred Credential Response identifying an issuance flow that contained one or more Credentials with the same Credential Configuration and Credential Dataset.

> * `event`: REQUIRED. Type of the notification event. It MUST be a case sensitive string whose value is either `credential_accepted`, `credential_failure`, or `credential_deleted`. `credential_accepted` is to be used when the Credentials were successfully stored in the Wallet, with or without user action. `credential_deleted` is to be used when the unsuccessful Credential issuance was caused by a user action. In all other unsuccessful cases, `credential_failure` is to be used. Partial errors during batch credential issuance (e.g., one of the Credentials in the batch could not be stored) MUST be treated as the overall issuance flow failing.

Reviewer 7: if the response is always a list containing a `credentials` key, does it make sense to rename these event to use the plural form like `credentials_accepted`...

   [[ To be removed from the final specification ]]

   -15

>    * credential response always returns an array when not returning a transaction_i with the option for additional meta-data

Reviewer 5: - There is a typo `transaction_i` instead of `transaction_id`
- I would add the following: 
  - `notification_id` moved to the `credentials` element 
  - Deferred Response returns `credentials` array, as well

Author: Agreed, should be fixed with my last commit

Reviewer 5: Thank you Christian

(That's the first section to check whenever I read the spec :smile: )

   [[ To be removed from the final specification ]]

   -15

   * credential response always returns an array when not returning a transaction_id with the option for additional meta-data
   * deferred credential response always returns an array (same as credential response)

> 

Reviewer 5: Until d14, `notification_id` was associated to a single credential.

I think that readers/implementers would appreciate to highlight the change on the granularity of `notification_id` (and notification events)

Author: That is a good point, I will add an explicit point to the document history
